### ISSUE TITLE
Fast filter with do block syntax

### ISSUE THREAD

**Issue by knuesel** [JuliaData/DataFrames.jl#2928]
The "slow" version of `filter` can be appealing for various reasons, I think one of them is that it enables the `do` block syntax. See [this discussion](https://discourse.julialang.org/t/dataframe-filtering-based-on-conditions-comparing-one-column-to-other-columns/70802/6) for an example where the preferred solution was using the slow filter:

```julia
using DataFrames

df = DataFrame(rand(8,5), :auto)

filter(df) do row
    all(<(row.x5), row[Not(Between(:x3, end))])
end
```

What about adding a fast method that enables `do` blocks? I'm thinking something like this:

```julia
Base.filter(fun, df::AbstractDataFrame, cols...; view::Bool=false) = filter(Cols(cols...) => fun, df; view)

filter(df, :x5, Not(Between(:x3, ncol(df)))) do x5, cols...
    all(<(x5), cols)
end
```

**Comment by bkamins** [comment]
This has been considered and we initially rejected it to avoid making API overly complicated.
An alternative to your proposal could be:
```
Base.filter(fun, df::AbstractDataFrame; cols=nothing, view::Bool=false) = filter(isnothing(cols) ? fun : cols => fun, df, view=view)
```
that would match `describe` API. If `cols` kwarg is `nothing` slow path is used.

CC @nalimilan

**Comment by knuesel** [comment]
Thinking about it a bit, I have a preference for the no-cols version, to me the `cols` argument seems an unnecessary complication in this case, so we have to write

```
filter(df, cols=Cols(:x5, Not(Between(:x3, ncol(df))))) do ...
```

instead of

```
filter(df, :x5, Not(Between(:x3, ncol(df)))) do ...
```

It makes sense for `describe` since the positional arguments are taken by the summary function. Also the semantics of these column selectors is a bit different: for `describe` the `cols` keyword (without transformation) means "do the same thing but on a subset of columns".  Calling `describe(df, cols=cols)` is basically equivalent to `describe(df[!, cols])`. So I would expect `filter(fun, df, cols=cols)` to call `fun` as in the slow path but passing `DataFrameRow`s that only contain the specified columns. I would find more natural for `filter` to have a syntax closer to `select`...

**Comment by bkamins** [comment]
> I would find more natural for `filter` to have a syntax closer to `select`

This is exactly my point. I do not want it to be close to `select` as it might confuse users (already it does) that these syntaxes are the same while they are not.

Let us wait for @nalimilan to comment.

**Comment by nalimilan** [comment]
I'm not sure making `filter` more powerful would be a good idea given that it's a bit special in terms of API and essentially exists for consistency with Base. If we wanted to support the `do` block syntax I think we should do that consistently across `subset`, `select`, `transform` and `combine` (and maybe others). I imagine that would be possible using `subset(cols, df) do ...`, which would be a shorthand for `subset(df, cols => ...)`.

But that would make the API even more complex than it is already. So for convenience I'd rather recommend using DataFramesMeta.

**Comment by bkamins** [comment]
Yes, this has started to pop-up recently https://discourse.julialang.org/t/why-does-filter-use-an-uncommon-position-for-the-input-dataframe/71004/10.

I will write a new docstring for `filter` anyway, see #2930, and thus we need to discuss if we want to "freeze it" or remove, or maybe add new features as proposed here.

**Comment by knuesel** [comment]
Ah I see how what I propose can be *more* confusing for some people. Maybe the syntax proposed by @nalimilan is less confusing? In the example that would still be a bit cumbersome:

```
filter(Cols(:x5, Not(Between(:x3, ncol(df)))), df) do x5, cols...
    all(<(x5), cols)
end
```

I liked in my original proposal that the `Cols` wrapper was not necessary, but maybe that's fine too.

**Comment by bkamins** [comment]
@knuesel - given the comment by @nalimilan I feel we are in the stage where we need to discuss (and convince ourselves) that we should keep adding features to `filter`. The general recommendation (and preference of the users) currently is to avoid using it - just keep it for backward compatibility reasons and recommend using `subset` .

In summary - let us first convince ourselves that `filter` is needed indeed. The point is that if we keep adding features to `filter` we will have to teach users:
something like use DataFrames.jl in such and such way which works across all functions except `filter` which is different and has a different (and not so easy) API.
